### ISSUE TITLE
Crash on startup due to memory issues, not fixed with |pack

### ISSUE THREAD

**Issue by deploradude** [urbit/urbit#3190]
**Describe the bug**
Receive this on startup, indicating full use of memory.  Only difference in recent days is importing ~400 entries of legacy internet 'blog.  Loaded once and used |pack as recommended, but then failed again.  Now bails every time.  Cannot run |mass as cannot access dojo.

urbit 0.10.8
boot: home is /home/span/urbit/docteg-mothep
loom: mapped 2048MB
lite: arvo formula 50147a8a
lite: core 590c9d56
lite: final state 590c9d56
loom: mapped 2048MB
boot: protected loom
live: loaded: GB/2.133.983.232
boot: installed 268 jets
---------------- playback starting ----------------
pier: replaying events 8752750-8754405
eyre: no channel to cancel ~[//http-server/0vr.gigd1/6/2]
pier: (8753250): play: done
pier: (8753751): play: done
|«play»
bail: meme
bailing out
pier: serf error: end of file

**To Reproduce**
Steps to reproduce the behaviour:
1. boot urbit ship after doing whatever changed the state.

**Expected behaviour**
Expect ship to boot without incident

**Screenshots**
n/a; see above for output

**System (please supply the following information, if relevant):**
 - OS: [e.g. macOS, linux64, FreeBSD]
 - 0.10.8
 - base hash: jp5fi

**Additional context**
Mirrored legacy internet 'blog to Publish.  Only recent difference.

**Notify maintainers**
n/a

**Comment by deploradude** [comment]
Ubuntu 20

**Comment by philipcmonk** [comment]
It's very strange that it looks like it's running out of memory while replaying events.  If they didn't run out of memory the first time, why would they the second time?  Maybe it's because we send several events at once, and it doesn't have room to store them?  cc @joemfb 

Can you run `./urbit-worker grab PIER`?  That runs |mass without having to actually start up, so hopefully that works.  Also try running `./urbit-worker pack PIER` and send the output of both of those.

**Comment by deploradude** [comment]
grab output:


`loom: mapped 2048MB
boot: protected loom
live: logical boot
boot: installed 268 jets
serf: measuring memory:
  warm jet state: B/288
  cold jet state: B/288
  hank cache: B/288
  battery hash cache: B/288
  hot jet state: KB/124.408
total jet stuff: KB/125.560
  bytecode cache: B/288
total nock stuff: B/288
  memoization cache: B/288
total road stuff: B/288
total marked: KB/126.136
free lists: KB/13.900
sweep: KB/126.136`

**Comment by deploradude** [comment]
pack:


`loom: mapped 2048MB
boot: protected loom
live: logical boot
boot: installed 268 jets
urbit-worker: pack: gained: KB/13.900`

**Comment by deploradude** [comment]
Sorry for formatting

**Comment by deploradude** [comment]
![image](https://user-images.githubusercontent.com/66261425/88569969-3f175c00-d009-11ea-82e6-8870d43ceac5.jpeg)

**Comment by deploradude** [comment]
Just kidding ...

![image](https://user-images.githubusercontent.com/66261425/88570467-20659500-d00a-11ea-9ced-460f23916838.jpeg)

**Comment by deploradude** [comment]
![image](https://user-images.githubusercontent.com/66261425/88570548-3a9f7300-d00a-11ea-9c25-cc37aa1527c2.jpeg)

**Comment by joemfb** [comment]
This is a *very* full loom, and already mostly compacted, so the |pack gains are minimal. Full deduplication would help here (ie, the previous |pack), but we don't yet have an implementation that can handle such a full memory arena.

I think @philipcmonk is right, this is failing because v0.10.8 batches replay events, and you're right up against the limit.

The basic recovery approach will be as follows: you'll most need to replay events in smaller batches, or even one at a time, packing anytime it fails. Once the whole log is replayed, you can run |mass to find out what's using so much memory. Then you can inject a `%trim` event to clear internal caches, or maybe take other evasive action.

Most of these steps are currently manual, and some require custom modifications to the binary. This is an area of active development, and still very immature. The quickest thing would be for you to send us a tarball of your pier (coordinating over support @ urbit.org) -- I can make the necessary modifications locally and recover the pier as much as I am able. Or you could wait a little longer, and I can replay back with links experimental binaries with new options enables for these various steps. (Actually, the absolute fastest fix is just for you to breach -- but that's a major pain, and you'd have to recreate groups and reinvite members.)

Let me know what you'd like to do. And sorry for the trouble -- getting control over these limitations (better tooling, warnings, &c) is my top priority.

**Comment by deploradude** [comment]
Thanks.  The pier should be sent to that address now.

**Comment by joemfb** [comment]
@deploradude, sorry for the delay. I have a solution for you. I've spent the last couple days working on the new tooling I referred to earlier, and have a working initial implementation. The branch is `jb/hashcons`, or you can download static binaries:

- [linux](https://bootstrap.urbit.org/b61c016b93b4b7f7e15b9a5a1da71f84698d0dd9-linux64.tgz)
- [macos](https://bootstrap.urbit.org/b61c016b93b4b7f7e15b9a5a1da71f84698d0dd9-darwin.tgz)

The new tool is currently available only in the urbit-worker binary. It works by globally reallocating and deduplicating your persistent state. It's run as follows:

```
./urbit-worker uniq path/to/pier
```

As this is experimental tooling, I recommend you backup of your snapshot first (the contents of `$pier/.urb/chk/`). (Note that these can be backed-up and restored at any point in time -- if they're stale, you'll just have to replay a portion of your event log. It's good practice to back them up periodically for long-lived or busy ships, as that could save significant replay time in the event of data corruption.

I'm confident in the tooling, despite its immaturity. We used it to save `~zod` earlier today. I also ran it on the pier you sent to the support address; the output was as follows:

```
$ time ./b61c016b93b4b7f7e15b9a5a1da71f84698d0dd9-linux64/urbit-worker uniq ./docteg-mothep
loom: mapped 2048MB
boot: protected loom
live: loaded: GB/2.133.983.232
boot: installed 268 jets
serf: measuring memory:
  kernel: GB/1.859.348.724
total arvo stuff: GB/1.859.348.724
  warm jet state: KB/112.532
  cold jet state: MB/272.259.948
  hank cache: B/288
  battery hash cache: B/288
  hot jet state: KB/124.468
total jet stuff: MB/272.497.524
  bytecode cache: B/288
total nock stuff: B/288
  trace buffer: B/36
  memoization cache: B/288
total road stuff: B/324
total marked: GB/2.131.846.860
free lists: KB/159.036
sweep: GB/2.131.846.860

hc: cells fill 0 size 233
hc: atoms fill 0 size 233
hc: cells fill 3930351 size 5702887
hc: atoms fill 351770 size 514229
hc: cold count 1223
hc: cells fill 8246742 size 9227465
hc: atoms fill 352442 size 514229
serf: measuring memory:
  kernel: MB/259.844.788
total arvo stuff: MB/259.844.788
  warm jet state: B/288
  cold jet state: MB/103.887.792
  hank cache: B/288
  battery hash cache: B/288
  hot jet state: B/288
total jet stuff: MB/103.888.944
  bytecode cache: B/288
total nock stuff: B/288
  memoization cache: B/288
total road stuff: B/288
total marked: MB/363.734.308
free lists: KB/49.596
sweep: MB/363.734.308


real    0m49.139s
user    0m37.325s
sys     0m9.220s
```

In other words, it took ~50s (on a basic gcp vm), and reduced your persistent state from ~2.1GB to ~360 MB. I could send you back the updated pier, but I think it'll be simpler for you to just run the command yourself.

Let me know if you have any questions, or run into any further issues.

**Comment by deploradude** [comment]
Joe, thank you

**Comment by deploradude** [comment]
Can I close this or do you guys need to?

**Comment by joemfb** [comment]
Let's leave it open 'till the fix is finished and integrated.

**Comment by ryjm** [comment]
Ran into this problem as well and `uniq` worked like a charm. Output:

```sh                                                                                   
loom: mapped 2048MB                                                                                                                          
boot: protected loom                                                                                                                         
live: loaded: GB/2.134.786.048                                                                                                               
boot: installed 268 jets                                                                                                                     
serf: measuring memory:                                                                                                                      
  kernel: GB/2.015.291.092                                                                                                                   
total arvo stuff: GB/2.015.291.092                                                                                                           
  warm jet state: KB/90.816                                                                                                                  
  cold jet state: MB/117.424.208                                                                                                             
  hank cache: B/288                                                                                                                          
  battery hash cache: B/288                                                                                                                  
  hot jet state: KB/124.720                                                                                                                  
total jet stuff: MB/117.640.320                                                                                                              
  bytecode cache: B/288                                                                                                                      
total nock stuff: B/288                                                                                                                      
  trace buffer: B/36                                                                                                                         
  memoization cache: B/288                                                                                                                   
total road stuff: B/324                                                                                                                      
total marked: GB/2.132.932.024                                                                                                               
free lists: KB/104.152                                                                                                                       
sweep: GB/2.132.932.024                                                                                                                      
                                                                                                                                             
hash-cons arena:                                                                                                                             
  root: B/144                                                                                                                                
  atoms (393561):                                                                                                                            
    refs: MB/10.284.580                                                                                                                      
    data: MB/179.188.524                                                                                                                     
    dict: MB/27.967.368                                                                                                                      
  total: MB/217.440.472                                                                                                                      
  cells (10066560):                                                                                                                          
    refs: MB/298.607.040                                                                                                                     
    dict: MB/812.016.920                                                                                                                     
  total: GB/1.110.623.960                                                                                                                    
total: GB/1.328.064.576                                                                                                                      
                                                                                                                                             
serf: measuring memory:                                                                                                                      
  kernel: MB/394.315.840                                                                                                                     
total arvo stuff: MB/394.315.840                                                                                                             
  warm jet state: KB/67.224                                                                                                                  
  cold jet state: MB/39.447.856                                                                                                              
  hank cache: B/288                                                                                                                          
  battery hash cache: B/288                                                                                                                  
  hot jet state: KB/124.408                                                                                                                  
total jet stuff: MB/39.640.064                                                                                                               
  bytecode cache: B/288                                                                                                                      
total nock stuff: B/288                                                                                                                      
  memoization cache: B/288                                                                                                                   
total road stuff: B/288                                                                                                                      
total marked: MB/433.956.480                                                                                                                 
free lists: KB/91.344                                                                                                                        
sweep: MB/433.956.480   
```

**Comment by ryjm** [comment]
Still runs out of memory eventually though, and I have to do another uniq.

**Comment by Splined** [comment]
I used this as well and it seemed to work the first few times.  Now urbit-worker itself will bail: oops.

$ ./urbit-worker uniq ./tinden-pontev
loom: mapped 2048MB
boot: protected loom
live: loaded: MB/625.147.904
boot: installed 268 jets
serf: measuring memory:
  kernel: MB/309.168.652
total arvo stuff: MB/309.168.652
  warm jet state: KB/90.848
  cold jet state: MB/224.292.896
  hank cache: B/288
  battery hash cache: B/288
  hot jet state: KB/124.408
total jet stuff: MB/224.508.728
  bytecode cache: B/288
total nock stuff: B/288
  memoization cache: B/288
total road stuff: B/288
total marked: MB/533.677.956
free lists: KB/381.764
sweep: MB/533.677.956

hash-cons arena:
  root: B/144
  atoms (440495):
    refs: MB/10.284.580
    data: MB/159.682.588
    dict: MB/27.967.368
  total: MB/197.934.536
  cells (15192933):
    refs: MB/483.156.340
    dict: GB/1.313.870.976
  total: GB/1.797.027.316
total: GB/1.994.961.996

serf: measuring memory:
  kernel: MB/309.168.652
total arvo stuff: MB/309.168.652
  warm jet state: KB/90.848
  cold jet state: MB/224.292.896
  hank cache: B/288
  battery hash cache: B/288
  hot jet state: KB/124.408
total jet stuff: MB/224.508.728
  bytecode cache: B/288
total nock stuff: B/288
  memoization cache: B/288
total road stuff: B/288
total marked: MB/533.677.956
free lists: KB/475.576
sweep: MB/533.677.956

Assertion '0' failed in noun/events.c:417

bail: oops
bailing out
Assertion failed: 0 (noun/manage.c: u3m_bail: 677)
Aborted (core dumped)

**Comment by joemfb** [comment]
The latest recovery in instructions are now here: https://github.com/urbit/urbit/issues/3645#issuecomment-709631355

**Comment by joemfb** [comment]
Closing in favor of #4182
